Edits to normalized cache article (#3631)
apollographql/apollo-android: docs/source/caching/normalized-cache.mdx

You can also use a normalized cache as a single source of truth for your UI, enabling it to react to changes in the cache. To learn more about the normalization process, [see this blog post](https://www.apollographql.com/blog/apollo-client/caching/demystifying-cache-normalization/).

### Normalizing responses

Look at this example query:

```graphql
query GetFavoriteBook {
  favoriteBook { # Book object
    id
    title
    author {     # Author object
      id
      name
    }
  }
}
```

This query returns a `Book` object, which in turn includes an `Author` object. An example response from the GraphQL server looks like this:

```json
{
  "favoriteBook": {
    "id": "bk123",
    "title": "Les Guerriers du silence",
    "author": {
      "id": "au456",
      "name": "Pierre Bordage"
    }
  }
}
```

A normalized cache does _not_ store this response directly. Instead, it breaks it up into the following entries by default:

> ```yaml:title=Cache

Reviewer: Nitpick
```suggestion
```json:title=Cache
```

Author: Cool, I wasn't sure whether these cache blocks were going to technically be valid _anything_ so I initially went with the most relaxed syntax 😄

> "favoriteBook": {"id": "bk123", "title": "Les guerriers du silence", "author": "CacheKey{favoriteBook.author}"

Reviewer 2: ```suggestion
"favoriteBook": {"id": "bk123", "title": "Les guerriers du silence", "author": "ApolloCacheReference{favoriteBook.author}"
```

"favoriteBook.author": {"id": "au456", name": "Pierre Bordage"}

> "QUERY_ROOT": {"favoriteBook": "CacheKey(favoriteBook)"}

Reviewer 2: ```suggestion
"QUERY_ROOT": {"favoriteBook": "ApolloCacheReference{favoriteBook}"}
```

```

> ⚠️ These default generated cache IDs (`favoriteBook` and `favoriteBook.author`) are undesirable for data deduplication. See [Specifying cache IDs](#specifying-cache-ids).

> * Notice that the `author` field of the `Book` entry now contains the string `CacheKey{favoriteBook.author}`. This is a **reference** to the `Author` cache entry.

Reviewer 2: ```suggestion
* Notice that the `author` field of the `Book` entry now contains the string `ApolloCacheReference{favoriteBook.author}`. This is a **reference** to the `Author` cache entry.
```

}
```

> Then include the SQLite cache in in your `ApolloClient` initialization **according to your platform target** (different platforms use different drivers):

Reviewer: Typo
```suggestion
Then include the SQLite cache in your `ApolloClient` initialization **according to your platform target** (different platforms use different drivers):
```

## Setting a fetch policy

After you add a normalized cache to your `ApolloClient` initialization, Apollo Android automatically uses it by default for all queries. You can customize _how_ the cache is used for a particular query by setting a **fetch policy** for that query.

The following snippet shows how to set all available fetch policies and their behavior:

```kotlin
val response = apolloClient.query(query)

  // (Default) Check the cache, then only use the network if data isn't present
  .fetchPolicy(FetchPolicy.CacheFirst)

  // Check the cache and never use the network, even if data isn't present
  .fetchPolicy(FetchPolicy.CacheOnly)

  // Always use the network, then check the cache if network fails
  .fetchPolicy(FetchPolicy.NetworkFirst)

  // Always use the network and never check the cache, even if network fails
  .fetchPolicy(FetchPolicy.NetworkOnly)

  // Execute the query
  .execute()
```

As with `normalizedCache(NormalizedCacheFactory)`, `fetchPolicy(FetchPolicy)` is an extension function on `ApolloClient.Builder()`, so you need `apollo-normalized-cache` in your classpath for this to work.

> Because the normalized cache deduplicates data, it enables you to react to cache changes. You do thiswith `watchers` that listen to cache changes. [Learn more about query watchers.](./query-watchers/)

Reviewer: ```suggestion
Because the normalized cache deduplicates data, it enables you to react to cache changes. You do this with `watchers` that listen to cache changes. [Learn more about query watchers.](./query-watchers/)
```

Reviewer: Very nitpick: should we say "observe" instead of listen? Mixing "watch" and "listen" may be odd :) Or not!

Reviewer 2: I'd say "listen" is more widespread at this point but that's completely subjective. Both work for me.

Author: Haha fair "observation"! Yeah I agree that "listen" is such a widespread word for this that it's fine 😄

> ## Specifying cache IDs

Reviewer 2: Any rationale behind the use of `object id` vs `cache id` vs `cache key`? I tend to prefer `cache key` everywhere but maybe these terms have slightly different meanings?

Author: I think we should avoid "Object ID" to reduce overlap with "an object's `id`", which might or not be the ID for the cache.

Between "cache ID" and "cache key", I went with "cache ID" purely because we already use that in the React docs and it seemed like a reasonable tie-breaker. I do think that "cache key" is a very reasonable alternative and maybe we should flip it for _every_ client in the future

By default, Apollo Android uses an object's GraphQL field path as its cache ID. For example, recall the following query and its resulting cache entries from earlier:

```graphql
query GetFavoriteBook {
  favoriteBook { # Book object
    id
    title
    author {     # Author object
      id
      name
    }
  }
}
```

> ```yaml:title=Cache

Reviewer: ```suggestion
```json:title=Cache
```

}
```

After executing this query, our cache looks like this:

> ```yaml{2-3}:title=Cache

Reviewer: ```suggestion
```json{2-3}:title=Cache
```

"favoriteBook": {"id": "bk123", "title": "Les guerriers du silence", "author": "CacheKey{favoriteBook.author}"
"favoriteBook.author": {"id": "au456", name": "Pierre Bordage"}
"author(\"id\": \"au456\")": {"id": "au456", "name": "Pierre Bordage"}
"QUERY_ROOT": {"favoriteBook": "CacheKey(favoriteBook)", "author(\"id\": \"au456\")": "CacheKey{author(\"id\": \"au456\")}"}
```

We're now caching two identical entries for the same `Author` object! This is undesirable for a few reasons:

* It takes up more space.
* Modifying one of these objects does _not_ notify any watchers of the _other_ object.

We want to deduplicate entries like these by making sure they're assigned the _same_ cache ID when they're written, resulting in a cache that looks more like this:

> ```yaml:title=Cache

Reviewer: ```suggestion
```json:title=Cache
```
